examples/chat.rs

```diff
@@ -0,0 +1,112 @@
+            // https://users.rust-lang.org/t/loop-futures-for-client-handling/6950/2
+            // First we need to get an infinite iterator
+            let iter = stream::iter::<_, _, std::io::Error>(iter::repeat(()).map(Ok));
+            // Then we fold it as infinite loop
+            let socket_reader = iter.fold(reader, move |reader, _| {
+                let connections = connections_inner.clone();
+                // read line
+                let amt = io::read_until(reader, '\n' as u8, vec![]);
+                // check if we hit EOF and need to close the connection
+                let amt = amt.and_then(|(reader, vec)| {
+                    // EOF was hit without reading a delimiter
+                    if vec.len() == 0 {
+                        let err = Error::new(ErrorKind::BrokenPipe, "Broken Pipe");
+                        futures::failed(err).boxed()
+                    } else {
+                        futures::finished((reader, vec)).boxed()
+                    }
+                });
+                // convert bytes into string
+                let amt = amt.map(|(reader, vec)| (reader, String::from_utf8(vec)));
+                amt.map(move |(reader, message)| {
+                    println!("{}: {:?}", addr, message);
+                    let conns = connections.borrow_mut();
+                    if let Ok(msg) = message {
+                        // For each open connection except the sender, send the string
+                        // via the channel
+                        let iter = conns.iter().filter(|&(&k,_)| k != addr).map(|(_,v)| v);
+                        for tx in iter {
+                            tx.send(format!("{}: {}", addr, msg)).unwrap();
+                        }
+                    } else {
+                        let tx = conns.get(&addr).unwrap();
+                        tx.send("You didn't send valid UTF-8.".to_string()).unwrap();
+                    }
+                    reader
+                })
+            });
+
+            // Whenever we receive a string on the Receiver, we write it to `WriteHalf<TcpStream>`.
+            let socket_writer = rx.fold(writer, |writer, msg| {
+                let amt = io::write_all(writer, msg.into_bytes());
+                let amt = amt.map(|(writer, _)| writer);
+                amt
+            });
+
+            // In order to fuse the reading and writing futures in the end, we need to have the
+            // same output type. As we don't need the values anymore, we can just map them
+            // to `()`.
+            let socket_reader = socket_reader.map(|_| ());
+            let socket_writer = socket_writer.map(|_| ());
+
+            let amt = socket_reader.select(socket_writer);
+            amt.then(move |_| {
+                connections.borrow_mut().remove(&addr);
+                println!("Connection {} closed.", addr);
+                Ok(())
+            })
+        };
+        handle.spawn_fn(main_fn);
```

Comment:
Could you please explain what you mean with a list of futures? An idea I have to reduce rightward drift is to wrap `connections` in a struct implementing all relevant functions. But this would make the code harder to understand. I'd like to keep the examples and their circular complexity as low as possible to make them understandable at first glance, which would not be the case with my idea. At the same time examples should show good code quality, which unfortunately not the case with the code because of rightward drift.